contracts/StorageManager.sol

```diff
@@ -46,48 +49,63 @@ contract StorageManager {
     event AgreementStopped(bytes32 indexed agreementReference);
 
+    constructor() Ownable() public { }
+
+    /**
+    @notice whitelist a token or remove the token from whitelist
+    @param token the token from whom you want to set the whitelisted
+    @param isWhiteListed whether you want to whitelist the token or put it from the whitelist.
+    */
+    function setWhitelistedTokens(address token, bool isWhiteListed) public onlyOwner {
+        isWhitelistedToken[token] = isWhiteListed;
+    }
+
+
     /**
     >> FOR PROVIDER
     @notice set the totalCapacity and billingPlans of a Offer.
     @dev
     - Use this function when initiating an Offer or when the users wants to change more than one parameter at once.
     - make sure that any period * prices does not cause an overflow, as this can never be accepted (REF_MAX_PRICE) and hence is pointless
+    - only whitelisted tokens are allowed to make a token for
+    - if there are two tokens, and two billingPrice/periods pairs per token, then boundaries[0] == 1. 
+      This makes the first two billingPeriod/prices pairs to apply to the first token ([tokens[0]]) and the second pairs to the second token.
     @param capacity the amount of bytes offered. If already active before and set to 0, existing contracts can't be prolonged / re-started, no new contracts can be started.
-    @param billingPeriods the offered periods. Length must be equal to the lenght of billingPrices.
-    @param billingPrices the prices for the offered periods. Each entry at index corresponds to the same index at periods. When a price is 0, the matching period is not offered.
+    @param billingPeriods the offered periods. Length must be equal to the length of billingPrices. The first index of the multi dem array corresponds with the address in tokens at the same index
+    @param billingPrices the prices for the offered periods. Each entry at index corresponds to the same index at periods. The first index of the multi dem array corresponds with the address in tokens at the same index
+    @param tokens the tokens for which an offer is made. By convention, address(0) is the native currency.
     @param message the Provider may include a message (e.g. his nodeID).  Message should be structured such that the first two bits specify the message type, followed with the message). 0x01 == nodeID
     */
     function setOffer(uint128 capacity,
-        uint64[] memory billingPeriods,
-        uint64[] memory billingPrices,
+        uint64[][] memory billingPeriods,
+        uint64[][] memory billingPrices,
+        address[] memory tokens,
         bytes32[] memory message
     ) public {
         Offer storage offer = offerRegistry[msg.sender];
         setTotalCapacity(capacity);
-        require(billingPeriods.length > 0, "StorageManager: Offer needs some billing plans");
-        require(billingPeriods.length == billingPrices.length, "StorageManager: Billing plans array length has to equal to billing prices");
```

Comment:
This check is already implicit because if not equal: `error: array index out of bounds`

Reply:
Not always true. Only true if one is longer than the other. Not an error if the other is longer than the one.

Reply:
Should make a comment about it if not checked here.

Reply:
I personally belief that Solidity should only contain verification, such that it ensures that the state is always consistent. Any business logic or seemingly surprising behavior for the user should be caught by the libraries or clients that interact with the contract. This is a good example of an unneeded check in Solidity code, as removing it doesn't risk bringing the contract in an invalid state. Nevertheless, we should add very explicitly to the `@dev` description on top of the function that billingPeriods length should be equal to billingPrices length.

Invalid input should then be caught by the client.

Reply:
TODO @Eknir , add this comment